### PR TITLE
[panasonictv] Initial Contribution

### PR DESCRIPTION
**Pull request by thinkingstone**
I would like to contribute/continue the work done for adding the panasonictv binding from oh1 to oh3. 
Therefore I used the existing migrations from #3904, #6179 and  #9024. 
There is also a discussion in https://community.openhab.org/t/panasonictv-oh3/108124

Issue for this PR #11559

### DESCRIPTION
- What is the classification of the PR? => new binding, oh1 migration
- Did you describe the PRs motivation and goal? => Yes
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread? => https://community.openhab.org/t/panasonictv-oh3/108124 and issues/PRs #3904, #6179 and  #9024
- Did you describe new features for the end user? => README.md
- Did you describe any noteworthy changes in usage for the end user? => No necessary
- Was the documentation updated accordingly, e.g. the add-on README? => Yes
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/guidelines.html => Yes
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/bindings/#include-the-binding-in-the-build => Yes
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing.html#sign-your-work => Yes

### CODE REVIEW

**Comment by Flole998** [comment]
As far as I can see you don't have the power state/command in there? Is there a specific reason for that?

**Comment by thinkingstone** [comment]
> As far as I can see you don't have the power state/command in there? Is there a specific reason for that?

You need to send the "Power" (or exactly "NRC_POWER-ONOFF") to the keycode channel. 
Or maybe I didn't understand what you mean. 

I didn't find a way to identify if the device is powered on. My tv can be in standby or powered on mode. In standby it can also be reachable by network or not, depending on it's network settings. If it is reachable the thing is considered online, but I can not determine if is powered on.

**Comment by Flole998** [comment]
I've described how that's done in #4407. Additionally you should only send the power key code then if the command state is != the current state, otherwise sending on would turn the TV off if it's already on.

File: bundles/org.openhab.binding.panasonictv/src/main/java/org/openhab/binding/panasonictv/internal/service/MediaRendererService.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                oldValue = "0";
+            }
+            newValue = DataConverters.convertCommandToIntValue(command, 0, 100, Integer.parseInt(oldValue));
+        } catch (NumberFormatException e) {
+            throw new NumberFormatException("Command '" + command + "' not supported");
+        }
+
+        updateResourceState(SERVICE_ID, "SetVolume",
+                Map.of("InstanceID", "0", "Channel", "Master", "DesiredVolume", Integer.toString(newValue)));
+        // updateResourceState(SERVICE_ID, "GetVolume", Map.of("InstanceID", "0", "Channel", "Master"));
```
Comment (fwolter) [inline bundles/org.openhab.binding.panasonictv/src/main/java/org/openhab/binding/panasonictv/internal/service/MediaRendererService.java:98]: Can this be removed? Same for below.

File: bundles/org.openhab.binding.panasonictv/src/main/resources/OH-INF/binding/binding.xml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<binding:binding id="panasonictv" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:binding="https://openhab.org/schemas/binding/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>Panasonic TV Binding</name>
+	<description>This is the binding for Panasonic TV. Binding should support all Panasonic Viera TV models</description>
+	<author>Prakashbabu Sidaraddi</author>
```
Comment (fwolter) [inline bundles/org.openhab.binding.panasonictv/src/main/resources/OH-INF/binding/binding.xml:8]: The author tag is deprecated and should therefore be removed. See openhab/openhab-core#1844.

File: bundles/org.openhab.binding.panasonictv/src/main/resources/OH-INF/thing/channel-types.xml [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<channel-type id="volume">
+		<item-type>Dimmer</item-type>
+		<label>Volume</label>
+		<description>Volume level of the TV.</description>
+		<category>SoundVolume</category>
+	</channel-type>
```
Comment (fwolter) [inline bundles/org.openhab.binding.panasonictv/src/main/resources/OH-INF/thing/channel-types.xml:12]: There's a predefined system channel you could use. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#system-state-channel-types

File: bundles/org.openhab.binding.panasonictv/src/main/resources/OH-INF/thing/channel-types.xml [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+	<channel-type id="mute">
+		<item-type>Switch</item-type>
+		<label>Mute</label>
+		<description>Mute state of the TV.</description>
+	</channel-type>
```
Comment (fwolter) [inline bundles/org.openhab.binding.panasonictv/src/main/resources/OH-INF/thing/channel-types.xml:18]: Same for mute.

File: bundles/org.openhab.binding.panasonictv/src/main/java/org/openhab/binding/panasonictv/internal/PanasonicTvBindingConstants.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
+/**
+ * Copyright (c) 2010-2021 Contributors to the openHAB project
```
Comment (fwolter) [inline bundles/org.openhab.binding.panasonictv/src/main/java/org/openhab/binding/panasonictv/internal/PanasonicTvBindingConstants.java:2]: This needs to be updated to 2022. You can do it automatically by invoking `mvn license:format` in the root dir of this repo.

**Review by fwolter — CHANGES_REQUESTED** [review]
Thanks for your contribution! I reviewed your code and here is my feedback.

There is one checkstyle warning left. You could take a look at `target/code-analysis/report.html` after building it with `mvn clean install`.

**Comment by thinkingstone** [comment]
> Thanks for your contribution! I reviewed your code and here is my feedback.

Thanks you very much. I will check your feedback. This may take some time.

File: bundles/org.openhab.binding.panasonictv/pom.xml [hunk 6]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>3.2.0-SNAPSHOT</version>
```
Comment (wborn) [inline bundles/org.openhab.binding.panasonictv/pom.xml:10]:
```suggestion
    <version>3.4.0-SNAPSHOT</version>
```

File: bundles/org.openhab.binding.panasonictv/pom.xml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
Comment (jlaur) [inline bundles/org.openhab.binding.panasonictv/pom.xml:3]:
```suggestion
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
```

See #11833

File: bundles/org.openhab.binding.panasonictv/pom.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (wborn) [reply bundles/org.openhab.binding.panasonictv/pom.xml:10]:
```suggestion
    <version>4.0.0-SNAPSHOT</version>
```

File: bundles/org.openhab.binding.panasonictv/src/main/resources/OH-INF/binding/binding.xml (hunk 2 above)
Comment (lsiepel) [reply bundles/org.openhab.binding.panasonictv/src/main/resources/OH-INF/binding/binding.xml:8]: This needs to be adapted to the new `addon.xml`, the PR to adapt the developer docs is not merged yet, so if you need more details, the corresponding issue is here: https://github.com/openhab/openhab-core/issues/2058
But you can also update your branch and look at another binding's `addon.xml` and adapt it to this initial contribution.

**Comment by lsiepel** [comment]
Unfortunately it looks like we cannot bring this PR to the finish line. After two years of inactivity i close this PR. If there is a need to restart, feel free to re-open or start a new PR.

**Pull request closed without merging**